lib/nxt_state_machine/state_machine.rb

```diff
@@ -0,0 +1,197 @@
+          event.state_machine.can_transition!(name, event.state_machine.current_state_name(self))
+          transition = event.event_transitions.resolve!(event.state_machine.current_state_name(self))
+          # Transition is build every time and thus should be thread safe!
+          transition.build_transition(event, self, set_state_method, *args, **opts)
+        end
+      end
+
+      class_context.define_method "can_#{name}?" do
+        event.state_machine.can_transition?(name, event.state_machine.current_state_name(self))
+      end
+    end
+
+    def can_transition?(event_name, from)
+      event = events.resolve!(event_name)
+      event && event.event_transitions.key?(from)
+    end
+
+    def can_transition!(event, from)
+      return true if can_transition?(event, from)
+      raise NxtStateMachine::Errors::TransitionNotDefined, "No transition :#{event} for state :#{from} defined"
+    end
+
+    def before_transition(from:, to:, run: nil, &block)
+      callbacks.register(from, to, :before, run, block)
+    end
+
+    def after_transition(from:, to:, run: nil, &block)
+      callbacks.register(from, to, :after, run, block)
+    end
+
+    def on_success(from:, to:, run: nil, &block)
+      callbacks.register(from, to, :success, run, block)
+    end
+
+    def defuse(errors = [], from:, to:)
+      defuse_registry.register(from, to, errors)
+    end
+
+    def on_error(error = StandardError, from:, to:, run: nil, &block)
+      error_callback_registry.register(from, to, error, run, block)
+    end
+
+    def on_error!(error = StandardError, from:, to:, run: nil, &block)
+      error_callback_registry.register!(from, to, error, run, block)
+    end
+
+    def around_transition(from:, to:, run: nil, &block)
+      callbacks.register(from, to, :around, run, block)
+    end
+
+    def configure(&block)
+      instance_exec(&block)
+      self
+    end
+
+    # TODO: Everything that require context should live in some sort of proxy
+    def run_before_callbacks(transition, context)
+      run_callbacks(transition, :before, context)
+    end
+
+    def run_after_callbacks(transition, context)
+      run_callbacks(transition, :after, context)
+    end
+
+    def run_success_callbacks(transition, context)
+      run_callbacks(transition, :success, context)
+    end
+
+    def find_error_callback(error, transition)
+      error_callback_registry.resolve(error, transition)
+    end
+
+    def run_callbacks(transition, kind, context)
+      current_callbacks = callbacks.resolve!(transition, kind)
+      return unless current_callbacks.any?
+
+      current_callbacks.each do |callback|
+        Callable.new(callback).bind(context).call(transition)
+      end
+
+      true
+    end
+
+    def current_state_name(context)
+      Callable.new(get_state_with).bind(context).call(target(context))
+    end
+
+    def target(context)
+      @target_method ||= (options[:target] || :itself)
+      context.send(@target_method)
+    end
+
+    private
+
```

Comment:
Public interface seems to be cluttered (a lot of public methods). Just an observation. Didn't really dig the purpose of each one, but worth to inspect and check if some of the could be made private.